src/dialer/dial-request.js

```diff
@@ -0,0 +1,197 @@
+'use strict'
+
+const AbortController = require('abort-controller')
+const AggregateError = require('aggregate-error')
+const pDefer = require('p-defer')
+const debug = require('debug')
+const log = debug('libp2p:dialer:request')
+log.error = debug('libp2p:dialer:request:error')
+const { AbortError } = require('libp2p-interfaces/src/transport/errors')
+
+const { anySignal } = require('../util')
+const { TokenHolder } = require('./token-holder')
+
+class DialRequest {
+  /**
+   *
+   * @param {object} options
+   * @param {Multiaddr[]} options.addrs
+   * @param {TransportManager} options.transportManager
+   * @param {Dialer} options.dialer
+   */
+  constructor ({
+    addrs,
+    dialAction,
+    dialer
+  }) {
+    this.addrs = addrs
+    this.dialer = dialer
+    this.dialAction = dialAction
+  }
+
+  /**
+   * @async
+   * @param {object} options
+   * @param {AbortSignal} options.signal An AbortController signal
+   * @param {number} options.timeout The max dial time for each request
+   * @returns {Connection}
+   */
+  async run (options) {
+    // Determine how many tokens we need
+    const tokensWanted = Math.min(this.addrs.length, this.dialer.perPeerLimit)
```

Comment:
The `dialer.getTokens` already checks `perPeerLimit`. 

I think that only the dialer should take care of this validation, as in the `dialRequest` perspective, it wants the the maximum possible and we may end up iterating in the future to "runtime" decisions, instead of a fixed limit